### PR TITLE
Fixed #34671 -- Inspectdb collation issue with oracle views

### PR DESCRIPTION
**Pull request by SecondPort**
This should solve the view problem, any recommendation is more than welcome

### CODE REVIEW

**Review by felixxm — COMMENTED** [review]
@SecondPort Thanks for the patch :+1: However, this is not an acceptable solution. First of all, it doesn't fix the issue for materialized views. Secondly, regression tests are missing. Also, it's complicated, and we should avoid building queries this way. The only thing that we need from `user_tables`/`users_views`/`users_mviews` is `default_collation`, I'd fetch the `default_collation` in the separate query (similar to the one used in `get_table_list()`) and pass it in `params` to the main query in `get_table_description()`. This way we can avoid JOINing with `user_tables`/`users_views`/`users_mviews`.

**Comment by SecondPort** [comment]
> @SecondPort Thanks for the patch 👍 However, this is not an acceptable solution. First of all, it doesn't fix the issue for materialized views. Secondly, regression tests are missing. Also, it's complicated, and we should avoid building queries this way. The only thing that we need from `user_tables`/`users_views`/`users_mviews` is `default_collation`, I'd fetch the `default_collation` in the separate query (similar to the one used in `get_table_list()`) and pass it in `params` to the main query in `get_table_description()`. This way we can avoid JOINing with `user_tables`/`users_views`/`users_mviews`.

Thank you very much for the answer and sorry for the mistake, I'll keep trying, have a good day. 😊

File: django/db/backends/oracle/introspection.py [hunk 1]
```diff
@@ -132,26 +143,26 @@ def get_table_description(self, cursor, table_name):
...
             FROM user_tab_cols
-            LEFT OUTER JOIN
-                user_tables ON user_tables.table_name = user_tab_cols.table_name
             LEFT OUTER JOIN
                 user_col_comments ON
-                user_col_comments.column_name = user_tab_cols.column_name AND
-                user_col_comments.table_name = user_tab_cols.table_name
+                    user_col_comments.column_name = user_tab_cols.column_name
+                    AND
+                    user_col_comments.table_name = user_tab_cols.table_name
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py:162]: Please revert unrelated indentation changes.

File: django/db/backends/oracle/introspection.py [hunk 2]
```diff
@@ -132,26 +143,26 @@ def get_table_description(self, cursor, table_name):
                 END as is_autofield,
                 CASE
                     WHEN EXISTS (
-                        SELECT  1
+                        SELECT 1
                         FROM user_json_columns
                         WHERE
-                            user_json_columns.table_name = user_tab_cols.table_name AND
+                            user_json_columns.table_name = user_tab_cols.table_name
+                            AND
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py:150]: Please revert unrelated indentation changes.

File: django/db/backends/oracle/introspection.py [hunk 3]
```diff
@@ -132,26 +143,26 @@ def get_table_description(self, cursor, table_name):
                 END as is_autofield,
                 CASE
                     WHEN EXISTS (
-                        SELECT  1
+                        SELECT 1
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py:146]: Please revert, this small space change should be handled separately.

File: tests/backends/oracle/test_introspection.py [hunk 4]
```diff
@@ -33,3 +33,13 @@ def test_get_sequences_manually_created_index(self):
...
+
+    def test_get_table_description(self):
+        with connection.cursor() as cursor:
+            cursor.execute("SELECT * FROM %s" % Square._meta.db_table)
+            description = connection.introspection.get_table_description(
+                cursor, Square._meta.db_table
+            )
+            self.assertEqual(len(description), 2)
+            self.assertEqual(description[0][0], "ID")
+            self.assertEqual(description[1][0], "NAME")
```
Comment (felixxm) [inline tests/backends/oracle/test_introspection.py:45]: This is not a proper regression test.

Comment (SecondPort) [reply tests/backends/oracle/test_introspection.py:45]: Regression tests that I should apply should be carried out in the tests.py file in the tests/admin_changelist path?

Comment (felixxm) [reply tests/backends/oracle/test_introspection.py]: Why in the `admin_changelist`? :thinking:  This bug has nothing to do with the admin.

Comment (SecondPort) [reply tests/backends/oracle/test_introspection.py]: I wanted to make sure I was understanding the change where it should be corrected, since in the vscode search engine I looked for regression tests and most of them appeared in that section, thank you very much for guiding me, have a good day

File: tests/backends/oracle/test_introspection.py [hunk 5]
```diff
@@ -33,3 +33,27 @@ def test_get_sequences_manually_created_index(self):
...
+                    cursor, "TEST_TABLE"
+                )
+
+                # Finally, we make assertions about the table description
+                self.assertEqual(len(description), 1)
+                self.assertEqual(description[0][0], "TEST_FIELD")
+                self.assertEqual(description[0][1], "VARCHAR2")
+            finally:
+                # Clean up after ourselves by removing the test table
+                cursor.execute("DROP TABLE TEST_TABLE")
```
Comment (felixxm) [inline tests/backends/oracle/test_introspection.py:59]: This ticket is about collations for view and materialized view, not for a table. Introspection for tables is already covered by `inspectdb.tests` and works fine, this is not a proper regression test. Do you have an Oracle test environment?

Comment (felixxm) [reply tests/backends/oracle/test_introspection.py:59]: This is an example of a regression test for this issue:
```diff
diff --git a/tests/backends/oracle/test_introspection.py b/tests/backends/oracle/test_introspection.py
index c6a143eb2d..e75d3a6c50 100644
--- a/tests/backends/oracle/test_introspection.py
+++ b/tests/backends/oracle/test_introspection.py
@@ -3,7 +3,7 @@ import unittest
 from django.db import connection
 from django.test import TransactionTestCase
 
-from ..models import Square
+from ..models import Person, Square
 
 
 @unittest.skipUnless(connection.vendor == "oracle", "Oracle tests")
@@ -33,3 +33,20 @@ class DatabaseSequenceTests(TransactionTestCase):
                 # Recreate model, because adding identity is impossible.
                 editor.delete_model(Square)
                 editor.create_model(Square)
+
+    def test_get_table_descriptioni_view_default_collation(self):
+        person_table = connection.introspection.identifier_converter(
+            Person._meta.db_table
+        )
+        person_view = connection.introspection.identifier_converter("TEST_PERSON_VIEW")
+        with connection.cursor() as cursor:
+            cursor.execute(f"CREATE VIEW {person_view} AS SELECT * FROM {person_table}")
+            try:
+                description = connection.introspection.get_table_description(
+                    cursor, person_view
+                )
+                self.assertEqual(len(description), 3)
+                for column in description:
+                    self.assertIsNone(column.collation)
+            finally:
+                cursor.execute(f"DROP VIEW {person_view}")
```

File: django/db/backends/oracle/introspection.py [hunk 6]
```diff
@@ -106,52 +106,29 @@ def get_table_list(self, cursor):
...
+
+        # Then we pass default_collation to the main query
+        cursor.execute(
+            """
+            SELECT column_name, data_type, data_default, nullable,
+                data_length, data_precision, data_scale, char_used,
+                char_length
+            FROM user_tab_columns
+            WHERE table_name = %s
+            ORDER BY column_id
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py:129]: Why you removed all the logic for other columns? (`is_json`, `is_autofield`, etc.) It seems that you don't have an Oracle setup.

File: django/db/backends/oracle/introspection.py [hunk 7]
```diff
@@ -106,52 +106,29 @@ def get_table_list(self, cursor):
...
-                user_tables ON user_tables.table_name = user_tab_cols.table_name
-            LEFT OUTER JOIN
-                user_col_comments ON
-                user_col_comments.column_name = user_tab_cols.column_name AND
-                user_col_comments.table_name = user_tab_cols.table_name
-            WHERE user_tab_cols.table_name = UPPER(%s)
-            """,
-            [table_name],
+            SELECT default_collation
+            FROM user_tables
```
Comment (felixxm) [inline django/db/backends/oracle/introspection.py:113]: Did you compare what is faster? `user_tables`, `user_tables` limited to the object types, or `UNION` of `user_tables`, `user_views`, and `user_mviews`.

**Comment by felixxm** [comment]
Superseded by #17012.
